inst/include/cpp11/r_vector.hpp

```diff
@@ -78,30 +78,77 @@ class r_vector {
 
   bool contains(const r_string& name) const;
 
+  // Same reasoning as `r_vector(const r_vector& rhs)` constructor
   r_vector& operator=(const r_vector& rhs) {
-    SEXP old_protect = protect_;
+    if (data_ == rhs.data_) {
+      return *this;
+    }
+
+    // Release existing object that we protect
+    detail::store::release(protect_);
 
     data_ = rhs.data_;
     protect_ = detail::store::insert(data_);
     is_altrep_ = rhs.is_altrep_;
     data_p_ = rhs.data_p_;
     length_ = rhs.length_;
 
-    detail::store::release(old_protect);
+    return *this;
+  };
+
+  // Same reasoning as `r_vector(r_vector&& rhs)` constructor
+  r_vector& operator=(r_vector&& rhs) {
+    if (data_ == rhs.data_) {
+      return *this;
+    }
+
+    // Release existing object that we protect
+    detail::store::release(protect_);
+
+    data_ = rhs.data_;
+    protect_ = rhs.protect_;
+    is_altrep_ = rhs.is_altrep_;
+    data_p_ = rhs.data_p_;
+    length_ = rhs.length_;
+
+    // Important for `rhs.protect_`, extra check for everything else
+    rhs.data_ = R_NilValue;
+    rhs.protect_ = R_NilValue;
+    rhs.is_altrep_ = false;
+    rhs.data_p_ = nullptr;
+    rhs.length_ = 0;
 
     return *this;
   };
 
+  // We are in read-only space so we can just copy over all properties except for
+  // `protect_`, which we need to manage on our own. `rhs` persists after this call, so we
+  // don't clear anything.
   r_vector(const r_vector& rhs) {
-    SEXP old_protect = protect_;
```

Comment:
This is a constructor, so `protect_` starts out as `R_NilValue`. Nothing to do here, no need to try and release